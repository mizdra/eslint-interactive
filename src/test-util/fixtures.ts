import { exec } from 'node:child_process';
import { randomUUID } from 'node:crypto';
import { rm } from 'node:fs/promises';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { promisify } from 'node:util';
import { defineIFFCreator } from '@mizdra/inline-fixture-files';
import fse from 'fs-extra/esm';

const execPromise = promisify(exec);

const rootDir = join(dirname(fileURLToPath(import.meta.url)), '..', '..');

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const fixtureDir = join(rootDir, 'tmp/fixture', process.env['VITEST_POOL_ID']!);
export const createIFF = defineIFFCreator({ generateRootDir: () => join(fixtureDir, randomUUID()) });

/**
 * Returns a string containing the stitched together contents of the file modified by fix.
 * To make the snapshot easier to read, the name of the file is inserted at the beginning of the contents of each file.
 */
export async function getSnapshotOfChangedFiles(): Promise<string> {
  const { stdout } = await execPromise(`diff -qr fixtures fixtures-tmp | cut -d " " -f 4 | xargs tail -n +1`, {
    cwd: rootDir,
  });
  return stdout.toString();
}

export async function setupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
  await fse.copy(resolve(rootDir, 'fixtures'), resolve(rootDir, 'fixtures-tmp'));
}

export async function cleanupFixturesCopy() {
  await rm(resolve(rootDir, 'fixtures-tmp'), { recursive: true, force: true });
}
