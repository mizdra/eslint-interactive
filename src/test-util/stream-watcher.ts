import { Readable } from 'node:stream';

export function createStreamWatcher(stream: Readable) {
  let allData = '';
  let unconsumedData = '';
  stream.on('data', (data) => {
    allData += data.toString();
    unconsumedData += data.toString();
  });
  async function waitOnData(): Promise<void> {
    return new Promise((resolve) => {
      stream.once('data', () => {
        resolve();
      });
    });
  }
  return {
    match: async (matcher: RegExp) => {
      while (!unconsumedData.match(matcher)) {
        // eslint-disable-next-line no-await-in-loop
        await waitOnData();
      }
      const ret = unconsumedData;
      unconsumedData = '';
      return ret;
    },
    getAllData() {
      return allData;
    },
  };
}
