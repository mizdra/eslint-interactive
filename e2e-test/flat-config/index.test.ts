import { afterEach, test, expect } from 'vitest';
import { createIFF } from '../../src/test-util/fixtures.js';
import dedent from 'dedent';
import { ChildProcessWithoutNullStreams, spawn } from 'child_process';
import { createStreamWatcher } from '../../src/test-util/stream-watcher.js';
import { readFile } from 'node:fs/promises';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

const ETX = String.fromCharCode(0x03); // ^C
const LF = String.fromCharCode(0x0a); // \n

const iff = await createIFF({
  'src/index.js': 'let a = 1;',
  'eslint.config.js': dedent`
    export default [
      { rules: { 'prefer-const': 'error' } },
    ];
  `,
  'package.json': '{ "type": "module" }',
});

function waitForClose(child: ChildProcessWithoutNullStreams) {
  return new Promise<void>((resolve) => {
    child.on('close', resolve);
  });
}

let child: ChildProcessWithoutNullStreams;
afterEach(async () => {
  child.kill();
  await waitForClose(child);
  await iff.reset();
});

test('fix problems with flat config', async () => {
  child = spawn(
    'node',
    [
      join(__dirname, '../../bin/eslint-interactive.js'),
      'src',
      // merge stderr to stdout
      '2>&1',
    ],
    { shell: true, stdio: 'pipe', cwd: iff.rootDir, env: { ...process.env, ESLINT_USE_FLAT_CONFIG: 'true' } },
  );
  const streamWatcher = createStreamWatcher(child.stdout, { debug: true });

  await streamWatcher.match(/Which rules would you like to apply action\?/);
  child.stdin.write(' '); // Select `prefer-const` rule
  child.stdin.write(LF); // Confirm the choice
  await streamWatcher.match(/Which action do you want to do\?/);
  child.stdin.write('1'); // Focus on `Run `eslint --fix``
  child.stdin.write(LF); // Confirm the choice
  await streamWatcher.match(/Fixing done\./);
  expect(await readFile(iff.paths['src/index.js'], 'utf-8')).toMatchSnapshot();
  child.stdin.write(ETX); // Exit
});
