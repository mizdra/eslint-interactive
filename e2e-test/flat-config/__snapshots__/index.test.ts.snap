// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fix problems with flat config 1`] = `"const a = 1;"`;
